File: DID.Swift.SDK/ElastosDIDSDKTests/VerifiableCredentialTest.swift

import XCTest
@testable import ElastosDIDSDK

class VerifiableCredentialTest: XCTestCase {
    
    func TestKycCredential() {
        do {
            let testData = TestData()
            
            // for integrity check
            _ = try testData.setupStore(true)
            let issuer:DIDDocument = try testData.loadTestIssuer()
            let test: DIDDocument = try testData.loadTestDocument()
            let vc: VerifiableCredential = try testData.loadEmailCredential()
            
            XCTAssertEqual(try DIDURL(test.subject, "email"), vc.getId())
            
            XCTAssertTrue((vc.getTypes()).contains("BasicProfileCredential"))
            XCTAssertTrue((vc.getTypes()).contains("InternetAccountCredential"))
            XCTAssertTrue((vc.getTypes()).contains("EmailCredential"))
            
            XCTAssertEqual(issuer.subject, vc.issuer)
            XCTAssertEqual(test.subject, vc.subject.did)
            XCTAssertEqual("john@example.com", vc.subject.getPropertyAsString(ofName: "email"))
            
            XCTAssertNotNil(vc.issuanceDate)
            XCTAssertNotNil(vc.expirationDate)
            
            XCTAssertFalse(vc.isExpired)
            XCTAssertTrue(vc.isGenuine)
            XCTAssertTrue(vc.isValid)
        }
        catch {
            XCTFail()
        }
    }
    
    func TestSelfProclaimedCredential() {
        do{
            let testData: TestData = TestData()
            
            // for integrity check
            _ = try testData.setupStore(true)
            let test: DIDDocument = try testData.loadTestDocument()
            
            let vc = try testData.loadProfileCredential()
            
            XCTAssertEqual(try DIDURL(test.subject, "profile"), vc!.getId())
            XCTAssertTrue((vc!.getTypes()).contains("BasicProfileCredential"))
            XCTAssertTrue((vc!.getTypes()).contains("SelfProclaimedCredential"))
            
            XCTAssertEqual(test.subject, vc!.issuer)
            XCTAssertEqual(test.subject, vc!.subject.did)
            
            XCTAssertEqual("John", vc!.subject.getPropertyAsString(ofName: "name"))
            XCTAssertEqual("Male", vc!.subject.getPropertyAsString(ofName: "gender"))
            XCTAssertEqual("Singapore", vc!.subject.getPropertyAsString(ofName: "nation"))
            XCTAssertEqual("English", vc!.subject.getPropertyAsString(ofName: "language"))
            XCTAssertEqual("john@example.com", vc!.subject.getPropertyAsString(ofName: "email"))
            XCTAssertEqual("@john", vc!.subject.getPropertyAsString(ofName: "twitter"))
            XCTAssertNotNil(vc!.issuanceDate)
            XCTAssertNotNil(vc!.expirationDate)
            
            XCTAssertFalse(vc!.isExpired)
            XCTAssertTrue(vc!.isGenuine)
            XCTAssertTrue(vc!.isValid)
        }
        catch {
            XCTFail()
        }
    }
    
    func testParseAndSerializeKycCredential() {
        do{
            let testData: TestData = TestData()
            
            var json: String = try testData.loadTwitterVcNormalizedJson()
            let normalized: VerifiableCredential = try VerifiableCredential.fromJson(json)
            
            json = try testData.loadTwitterVcCompactJson()
            let compact: VerifiableCredential = try VerifiableCredential.fromJson(json)
            
            let vc: VerifiableCredential = try testData.loadTwitterCredential()
            
            XCTAssertEqual(try testData.loadTwitterVcNormalizedJson(), normalized.toString(true))
            XCTAssertEqual(try testData.loadTwitterVcNormalizedJson(), compact.toString(true))
            XCTAssertEqual(try testData.loadTwitterVcNormalizedJson(), vc.toString(true))
            
            XCTAssertEqual(try testData.loadTwitterVcCompactJson(), normalized.toString(false))
            XCTAssertEqual(try testData.loadTwitterVcCompactJson(), compact.toString(false))
            XCTAssertEqual(try testData.loadTwitterVcCompactJson(), vc.toString(false))
        }
        catch {
         print(error)
            XCTFail()
        }
    }
    
    func testParseAndSerializeSelfProclaimedCredential() {
        do {
            let testData: TestData = TestData()
            
            var json: String = try testData.loadProfileVcNormalizedJson()
            let normalized: VerifiableCredential = try VerifiableCredential.fromJson(json)
            
            json = try testData.loadProfileVcCompactJson()
            let compact: VerifiableCredential = try VerifiableCredential.fromJson(json)
            
            let vc = try testData.loadProfileCredential()
            
            XCTAssertEqual(try testData.loadProfileVcNormalizedJson(), normalized.toString(true))
            XCTAssertEqual(try testData.loadProfileVcNormalizedJson(), compact.toString(true))
            XCTAssertEqual(try testData.loadProfileVcNormalizedJson(), vc!.toString(true))
            
            XCTAssertEqual(try testData.loadProfileVcCompactJson(), normalized.toString(false))
            XCTAssertEqual(try testData.loadProfileVcCompactJson(), compact.toString(false))
            XCTAssertEqual(try testData.loadProfileVcCompactJson(), vc!.toString(false))
        }
        catch {
            XCTFail()
        }
    }
    
    func testParseAndSerializeJsonCredential() {
        do {
            let testData = TestData()
            var json = try testData.loadJsonVcNormalizedJson()
            let normalized = try VerifiableCredential.fromJson(json)
            json = try testData.loadJsonVcCompactJson()
            let compact = try VerifiableCredential.fromJson(json)
            let vc = try testData.loadJsonCredential()
            XCTAssertEqual(try testData.loadJsonVcNormalizedJson(), normalized.toString(true))
            XCTAssertEqual(try testData.loadJsonVcNormalizedJson(), compact.toString(true))
            XCTAssertEqual(try testData.loadJsonVcNormalizedJson(), vc.toString(true))

            XCTAssertEqual(try testData.loadJsonVcCompactJson(), normalized.toString(false))
            XCTAssertEqual(try testData.loadJsonVcCompactJson(), compact.toString(false))
            XCTAssertEqual(try testData.loadJsonVcCompactJson(), vc.toString(false))
        } catch {
            XCTFail()
        }
    }
}

File: DID.Java.SDK/src/test/java/org/elastos/did/VerifiableCredentialTest.java
/*
 * Copyright (c) 2019 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.elastos.did;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

import java.io.IOException;
import java.util.Arrays;

import org.elastos.did.exception.DIDException;
import org.junit.Test;

public class VerifiableCredentialTest {
	@Test
	public void TestKycCredential() throws DIDException, IOException {
		TestData testData = new TestData();

		// for integrity check
		testData.setup(true);
		DIDDocument issuer = testData.loadTestIssuer();
		DIDDocument test = testData.loadTestDocument();

		VerifiableCredential vc = testData.loadEmailCredential();

		assertEquals(new DIDURL(test.getSubject(), "email"), vc.getId());

		assertTrue(Arrays.asList(vc.getTypes()).contains("BasicProfileCredential"));
		assertTrue(Arrays.asList(vc.getTypes()).contains("InternetAccountCredential"));
		assertTrue(Arrays.asList(vc.getTypes()).contains("EmailCredential"));

		assertEquals(issuer.getSubject(), vc.getIssuer());
		assertEquals(test.getSubject(), vc.getSubject().getId());

		assertEquals("john@example.com", vc.getSubject().getPropertyAsString("email"));

		assertNotNull(vc.getIssuanceDate());
		assertNotNull(vc.getExpirationDate());

		assertFalse(vc.isExpired());
		assertTrue(vc.isGenuine());
		assertTrue(vc.isValid());
	}

	@Test
	public void TestSelfProclaimedCredential() throws DIDException, IOException {
		TestData testData = new TestData();

		// for integrity check
		testData.setup(true);
		DIDDocument test = testData.loadTestDocument();

		VerifiableCredential vc = testData.loadProfileCredential();

		assertEquals(new DIDURL(test.getSubject(), "profile"), vc.getId());

		assertTrue(Arrays.asList(vc.getTypes()).contains("BasicProfileCredential"));
		assertTrue(Arrays.asList(vc.getTypes()).contains("SelfProclaimedCredential"));

		assertEquals(test.getSubject(), vc.getIssuer());
		assertEquals(test.getSubject(), vc.getSubject().getId());

		assertEquals("John", vc.getSubject().getPropertyAsString("name"));
		assertEquals("Male", vc.getSubject().getPropertyAsString("gender"));
		assertEquals("Singapore", vc.getSubject().getPropertyAsString("nation"));
		assertEquals("English", vc.getSubject().getPropertyAsString("language"));
		assertEquals("john@example.com", vc.getSubject().getPropertyAsString("email"));
		assertEquals("@john", vc.getSubject().getPropertyAsString("twitter"));

		assertNotNull(vc.getIssuanceDate());
		assertNotNull(vc.getExpirationDate());

		assertFalse(vc.isExpired());
		assertTrue(vc.isGenuine());
		assertTrue(vc.isValid());
	}

	@Test
	public void testParseAndSerializeKycCredential()
			throws DIDException, IOException {
		TestData testData = new TestData();

		String json = testData.loadTwitterVcNormalizedJson();
		VerifiableCredential normalized = VerifiableCredential.fromJson(json);

		json = testData.loadTwitterVcCompactJson();
		VerifiableCredential compact = VerifiableCredential.fromJson(json);

		VerifiableCredential vc = testData.loadTwitterCredential();

		assertEquals(testData.loadTwitterVcNormalizedJson(), normalized.toString(true));
		assertEquals(testData.loadTwitterVcNormalizedJson(), compact.toString(true));
		assertEquals(testData.loadTwitterVcNormalizedJson(), vc.toString(true));

		assertEquals(testData.loadTwitterVcCompactJson(), normalized.toString(false));
		assertEquals(testData.loadTwitterVcCompactJson(), compact.toString(false));
		assertEquals(testData.loadTwitterVcCompactJson(), vc.toString(false));
	}

	@Test
	public void testParseAndSerializeSelfProclaimedCredential()
			throws DIDException, IOException {
		TestData testData = new TestData();

		String json = testData.loadProfileVcNormalizedJson();
		VerifiableCredential normalized = VerifiableCredential.fromJson(json);

		json = testData.loadProfileVcCompactJson();
		VerifiableCredential compact = VerifiableCredential.fromJson(json);

		VerifiableCredential vc = testData.loadProfileCredential();

		assertEquals(testData.loadProfileVcNormalizedJson(), normalized.toString(true));
		assertEquals(testData.loadProfileVcNormalizedJson(), compact.toString(true));
		assertEquals(testData.loadProfileVcNormalizedJson(), vc.toString(true));

		assertEquals(testData.loadProfileVcCompactJson(), normalized.toString(false));
		assertEquals(testData.loadProfileVcCompactJson(), compact.toString(false));
		assertEquals(testData.loadProfileVcCompactJson(), vc.toString(false));
	}

	@Test
	public void testParseAndSerializeJsonCredential()
			throws DIDException, IOException {
		TestData testData = new TestData();

		String json = testData.loadJsonVcNormalizedJson();
		VerifiableCredential normalized = VerifiableCredential.fromJson(json);

		json = testData.loadJsonVcCompactJson();
		VerifiableCredential compact = VerifiableCredential.fromJson(json);

		VerifiableCredential vc = testData.loadJsonCredential();

		assertEquals(testData.loadJsonVcNormalizedJson(), normalized.toString(true));
		assertEquals(testData.loadJsonVcNormalizedJson(), compact.toString(true));
		assertEquals(testData.loadJsonVcNormalizedJson(), vc.toString(true));

		assertEquals(testData.loadJsonVcCompactJson(), normalized.toString(false));
		assertEquals(testData.loadJsonVcCompactJson(), compact.toString(false));
		assertEquals(testData.loadJsonVcCompactJson(), vc.toString(false));
	}
}
